core/metadata_vec.h

```diff
@@ -0,0 +1,631 @@
+     */
+    std::vector<MDObject> getObjectsForActiveLabels() const;
+
+
+    /** This two variables will be used to read the metadata information (labels and size)
+     * or maybe a few rows only
+     */
+    size_t _maxRows, _parsedLines;
+
+public:
+    /** @name Constructors
+     *  @{
+     */
+
+    /** Empty Constructor.
+     *
+     * The MetaData is created with no data stored on it. You can fill in it programmatically
+     * or by a later reading from a MetaData file or old Xmipp formatted type.
+     * if labels vectors is passed this labels are created on metadata
+     */
+    MetaDataVec();
+    MetaDataVec(const std::vector<MDLabel> &labelsVector);
+
+    /** From File Constructor.
+     *
+     * The MetaData is created and data is read from provided FileName. Optionally, a vector
+     * of labels can be provided to read just those required labels
+     */
+    MetaDataVec(const FileName &fileName, const std::vector<MDLabel> &desiredLabels);
+    MetaDataVec(const FileName &fileName);
+
+    /** Copy constructor
+     *
+     * Created a new metadata by copying all data from an existing MetaData object.
+     */
+    MetaDataVec(const MetaData &md);
+    MetaDataVec(const MetaDataVec &md) = default;
+
+    /** Assignment operator
+     *
+     * Copies MetaData from an existing MetaData object.
+     */
+    MetaDataVec& operator=(const MetaData &md);
+    MetaDataVec& operator=(const MetaDataVec &md) = default;
+
+    /**Clear all data
+     */
+    void clear() override;
+    /** @} */
+
+    /** @name Getters and setters
+     * @{
+     */
+
+    bool nextBlock(mdBuffer &buffer, mdBlock &block);
+
+    /** Export medatada to xml file.
+     *
+     */
+    void writeXML(const FileName fn, const FileName blockname, WriteModeMetaData mode) const override;
+
+    /** Write metadata in text file as plain data without header.
+     *
+     */
+    void writeText(const FileName fn,  const std::vector<MDLabel>* desiredLabels) const override;
+
+    void _parseObjects(std::istream &is, std::vector<MDObject*> & columnValues, const std::vector<MDLabel> *desiredLabels, bool firstTime);
+
+    /* Helper function to parse an MDObject and set its value.
+     * The parsing will be from an input stream(istream)
+     * and if parsing fails, an error will be raised
+     */
+    void _parseObject(std::istream &is, MDObject &object, size_t id = BAD_OBJID);
+
+    /** Get Metadata labels for the block defined by start
+     * and end loop pointers. Return pointer to newline after last label
+     */
+    void _readColumnsStar(mdBlock &block,
+                          std::vector<MDObject*> & columnValues,
+                          const std::vector<MDLabel>* desiredLabels,
+                          bool addColumns = true,
+                          size_t id = BAD_OBJID);
+
+    /** @} */
+
+    /** @name MetaData Manipulation
+     * @{
+     */
+
+    template <typename T> // T could be only MDRow or MDRowConst
```

Comment:
T can be compile-time checked via type trails

Reply:
Outdated.